lib/inputs/dovetail-frequency.js

```diff
@@ -0,0 +1,159 @@
+    const [successCount, deleteCount, failureCount] = await this.updateAll(updates);
+    const output = [];
+    output.push({ dest: 'updates', count: successCount });
+    output.push({ dest: 'deletes', count: deleteCount });
+
+    // just throw an error if anything failed. successful records have already
+    // been recorded in DDB, so retrying them won't cause duplicates.
+    if (failureCount) {
+      const msg = `DovetailFrequency retrying for ${failureCount}/${failureCount + successCount}`;
+      logger.warn(msg, { ddb: 'retrying', count: failureCount });
+      const err = new Error(msg);
+      err.skipLogging = true;
+      throw err;
+    }
+
+    return output;
+  }
+
+  // spin up N workers, and process the queue of updates
+  async updateAll(updates, concurrency = 25) {
+    let successCount = 0;
+    let deleteCount = 0;
+    let failureCount = 0;
+
+    const client = await dynamo.client();
+    const worker = async () => {
+      let args;
+      while ((args = updates.shift())) {
+        try {
+          const params = this.updateParams(args);
+          const result = await dynamo.updateItemPromise(params, client);
+
+          const toDelete = this.evictImpressions(result);
+          if (toDelete.length > 0) {
+            const deleteParams = this.deleteParams(args, toDelete);
+            const deleteResult = await dynamo.updateItemPromise(deleteParams, client);
+            deleteCount++;
+          }
+
+          successCount++;
+        } catch (err) {
+          if (err.name === 'ProvisionedThroughputExceededException') {
+            logger.warn(`DDB throughput exceeded [${process.env.DDB_TABLE}]: ${err}`, { args });
+          } else {
+            logger.error(`DDB Error [${process.env.DDB_TABLE}]: ${err}`, { args });
+          }
+          failureCount++;
+        }
+      }
+    };
+
+    const threads = Array(concurrency).fill(true);
+    await Promise.all(threads.map(() => worker()));
+    return [successCount, deleteCount, failureCount];
+  }
+
+  evictImpressions(result) {
+    if (result.Attributes && result.Attributes.impressions) {
+      const evict = Math.floor((new Date().getTime() / 1000) - FREQUENCY_HORIZON_SECONDS);
+      return result.Attributes.impressions.NS.filter((imp) => imp <= evict);
+    }
+    return [];
+  }
+
+  deleteParams(data, deleteArray) {
+    return {
+      TableName: this.tableName(),
+      Key: {
+        listener: { S: `${data.listener}` },
+        campaign: { S: `${data.campaign}` },
+      },
+      ReturnValues: 'ALL_NEW',
+      UpdateExpression: "DELETE impressions :td",
+      ExpressionAttributeValues: {
+        ":td": { "NS": deleteArray.map((td) => `${td}`) },
+      },
+    };
+  }
+
+  updateParams(data) {
+    return {
+      TableName: this.tableName(),
+      Key: {
+        listener: { S: `${data.listener}` },
+        campaign: { S: `${data.campaign}` },
+      },
+      ReturnValues: 'ALL_OLD',
```

Comment:
You want the old values, not the new/current?

Reply:
I use the old values to know what old impression values I can delete if they are past the horizon, without having to make another request to get them.